Update First.cpp

SobolevNick opened the pull request.

SobolevNick committed:
Update First.cpp

DimaTrushin commented on First.cpp:
```diff
@@ -1,81 +1,85 @@
+  using Vector = Eigen::VectorXf;
+  using RowVector = Eigen::RowVectorXf;
+  NetLayer(int m, int n)
+      : A_(Matrix::Random(m,n)), b_(Vector::Random(m)) {
```
пробле после запятой в функции Random.

DimaTrushin commented on First.cpp:
```diff
@@ -1,81 +1,85 @@
+  Matrix count_grad_A(Vector x, Vector u) const {
+    Matrix S(out_size(),out_size());
+    for (int i = 0; i < out_size(); i++) {
+      S(i,i) = sigma::evaluate1((A_ * x + b_)(i));
```
- Вот тут в каждой итерации цикла ты рассчитываешь A_ * x + b_. Лучше его предрасчитать до цикла.
- На самом деле у Eigen есть способ делать диагональные матрицы из вектора 
http://eigen.tuxfamily.org/index.php?title=SpecialMatrix#Diagonal_matrix
Ты можешь использовать этот метод для имплементации evaluate1, чтобы она возвращала именно диагональную матрицу.

DimaTrushin commented on First.cpp:
```diff
@@ -1,81 +1,85 @@
 };
+}
+void test_net_layer() {
+  ML::NetLayer print_weights(); //как-то не работает 
```
А это и не должно работать. По стандарту тут у тебя определяется функция с именем print_weights. Если хочешь тестировать свою класс, то надо сначала создать элемент класса, например
```cpp
ML::NetLayer Layer1(20, 30);
Layer1.print_weights();
ML::NetLayer Layer2(25, 35);
Layer2.print_bias();
```

DimaTrushin commented on First.cpp:
```diff
@@ -1,81 +1,85 @@
+    return x.array().exp() / (1 + x.array().exp());
+  };
+  //static Eigen::MatrixXf evaluate1(const Eigen::VectorXf& x) {
+    //return Eigen::MatrixXf::Identity().array() * (x.array().exp() / pow(1 + x.array().exp(), 2));
```
Думаю можно имплементировать как-то так
```cpp
return (x.array().exp() / pow(1 + x.array().exp(), 2)).matrix().asDiagonal();
```
Покопайся в документации, если я был не совсем точен, найди рабочую версию.

DimaTrushin commented on First.cpp:
```diff
@@ -1,81 +1,85 @@
+  }
+  Matrix count_grad_A(Vector x, Vector u) const {
+    Matrix S(out_size(),out_size());
+    for (int i = 0; i < out_size(); i++) {
```
Не используй i++, старайся всегда использовать преинкремент ++i. Для типа int это почти не влияет на производительность и ты не заметишь разницы, плюс компилятор может оптимизировать правильно. Но для сложных итераторов, это может быть критично.

DimaTrushin commented on First.cpp:
```diff
@@ -1,81 +1,85 @@
+    return sigma::evaluate0(A_ * x + b_);
+  }
+  Matrix count_grad_A(Vector x, Vector u) const {
+    Matrix S(out_size(),out_size());
```
пробел после запятой.

DimaTrushin commented on First.cpp:
```diff
@@ -1,81 +1,85 @@
+  Vector count_grad_b(Vector x, Vector u) const {
+    Matrix S(out_size(),out_size());
+    for (int i = 0; i < out_size(); i++) {
+      S(i,i) = sigma::evaluate1((A_ * x + b_)(i));
```
пробел после запятой

DimaTrushin commented on First.cpp:
```diff
@@ -1,81 +1,85 @@
+
+namespace ML {
+class sigma {
 public:
```
Здесь тоже можно добавить using-ги Matrix и Vector для удобства.

DimaTrushin reviewed (changes requested):
Сделал еще несколько замечаний. Сделал предложение как имплементировать матричную версию для evaluate1 в sigma.

SobolevNick committed:
Update First.cpp

DimaTrushin commented on First.cpp:
```diff
@@ -1,85 +1,90 @@
+        Vector predict(const Vector &x) const {
+            return sigma::evaluate0(A_ * x + b_);
+        }
+        Matrix count_grad_A(Vector x, Vector u) const {
```
Аргументы по константной ссылке.

DimaTrushin commented on First.cpp:
```diff
@@ -1,85 +1,90 @@
+        Matrix count_grad_A(Vector x, Vector u) const {
+            return sigma::evaluate1(A_ * x + b_) * u * x.transpose();
+        }
+        Vector count_grad_b(Vector x, Vector u) const {
```
Аргументы по константной ссылке.

DimaTrushin commented on First.cpp:
```diff
@@ -1,85 +1,90 @@
+        Vector count_grad_b(Vector x, Vector u) const {
+            return sigma::evaluate1(A_ * x + b_) * u;
+        }
+        Vector count_grad_x(Vector x, Vector u) const {
```
Аргументы по константной ссылке.

SobolevNick committed:
Update First.cpp

DimaTrushin commented on First.cpp:
```diff
@@ -25,39 +28,74 @@ namespace ML {
+    class NeuralNetwork {
+        using Matrix = Eigen::MatrixXf;
+        using Vector = Eigen::VectorXf;
+        NeuralNetwork(Vector x, Vector y) {
```
Аргументы по константной ссылке, иначе лишние копирования.

SobolevNick replied:
Поправил

DimaTrushin commented on First.cpp:
```diff
@@ -25,39 +28,74 @@ namespace ML {
+        using Matrix = Eigen::MatrixXf;
+        using Vector = Eigen::VectorXf;
+        NeuralNetwork(Vector x, Vector y) {
+            NetLayer Layer((y.rows()), (x.rows()));
```
В этом месте ты создаешь локальную переменную в конструкторе и потом ее уничтожаешь, при выходе из конструктора. Это ненужное действие.

DimaTrushin commented on First.cpp:
```diff
@@ -25,39 +28,74 @@ namespace ML {
+    class NeuralNetwork {
+        using Matrix = Eigen::MatrixXf;
+        using Vector = Eigen::VectorXf;
+        NeuralNetwork(Vector x, Vector y) {
```
В целом конструктор очень странный. Я бы ожидал увидеть в NeuralNetwork следующие переменные
```cpp
NetLayer Layer1_;
NetLayer Layer2_;
LossFunction LF_;
```
При этом LossFunction у тебя сейчас по сути не  настраиваемая и ее можно не хранирть тут. Но я ее помещаю в нейросеть для того, чтобы в будущем была поддержка настраиваемости.
Далее в конструкторе я бы ожидал увидеть следующую сигнатуру
```cpp
NeuralNetwork(int size1, int size2, int size3) : Layer1_(size2, size1), Layer2_(size3, size2) {}
```

DimaTrushin commented on First.cpp:
```diff
@@ -25,39 +28,74 @@ namespace ML {
+        NeuralNetwork(Vector x, Vector y) {
+            NetLayer Layer((y.rows()), (x.rows()));
+        }
+        static std::pair<Matrix, Vector> count_theta_2(Matrix A, Vector b) {
```
В публичном иинтерфейсе у нейросети должны быть два метода
```cpp
struct TrainingDatum {
  Vector Input;
  Vector output;
};
using TrainingData = std::vector<TrainingDatum>;

void train(const TrainingData& d);
Vector predict(const Vector&  x) const;
```
Первый метод делает обучение нейросети по данным обучающей выборки.
Второй метод считает предстказание сети по входу.

DimaTrushin commented on First.cpp:
```diff
@@ -25,39 +28,74 @@ namespace ML {
+        };
+
+    private:
+        Vector z_1;
```
Если я правильно понял назначение этого вектора -- это буффер для сохранения промежуточных вычислений. Его не надо помещать в нейросеть. Эти буферы должны создаваться в функции train только во время тренировки сети. Иначе ты используешь лишнюю память все время.

DimaTrushin commented on First.cpp:
```diff
@@ -25,39 +28,74 @@ namespace ML {
+    class LossFunction {
+    public:
+        using Vector = Eigen::VectorXf;
+        static float count_loss_function(std::vector<std::pair<Eigen::VectorXf, Eigen::VectorXf>> x) {
```
Вместо вот этого безобразия в виде вектора от пар векторов, используй using-ги так
```cpp
struct TrainingDatum {
  Vector Input;
  Vector output;
};
using TrainingData = std::vector<TrainingDatum>;
```
и тогда ты сможешь написать
```cpp
static float count_loss_function(const TrainingData& data);
```

SobolevNick replied:
Сделано

DimaTrushin commented on First.cpp:
```diff
@@ -25,39 +28,74 @@ namespace ML {
+    public:
+        using Vector = Eigen::VectorXf;
+        static float count_loss_function(std::vector<std::pair<Eigen::VectorXf, Eigen::VectorXf>> x) {
+            NetLayer Layer(x[0].first.size(), x[0].second.size());
```
- Во-первых, этот код упадет на пустом векторе обучающих данных.
- Во-вторых, какого хрена функция рассчета loss function у тебя создает слой сети?

DimaTrushin commented on First.cpp:
```diff
@@ -25,39 +28,74 @@ namespace ML {
+        static float count_loss_function(std::vector<std::pair<Eigen::VectorXf, Eigen::VectorXf>> x) {
+            NetLayer Layer(x[0].first.size(), x[0].second.size());
+            float sum = 0;
+            for (int i = 0; i < x.size(); i++) {
```
У тебя не правильная логика работы функции.
- Для начала сделай простую функцию рассчета. А именно. Она должна зависеть от трех параметров (x, y) -- обучающая пара и z -- значение выхода нейросети на x. Тогда ты на вход должен принять (x, y, z) и по ним построить значение loss function.
- Второй этап сделать loss function по нескольким обучающим данным. На вход тебе подаются пары (x1,y1),...,(xn,yn) и вычисления z1,...,zn на входах x1,...,xn. И тут просто вызвать предыдущую функцию n раз и усреднить результат. Хотя на мой взгляд, этот этап лучше вообще не делать. И эту логику реализовать в самой нейросети в результате ее обучения.

DimaTrushin commented on First.cpp:
```diff
@@ -25,39 +28,74 @@ namespace ML {
+            }
+            return sum / x.size();
+        }
+        static Vector count_starting_gradient(std::vector<std::pair<Eigen::VectorXf, Eigen::VectorXf>> x) {
```
Все замечания по поводу предыдущей функции актуальны и тут. У тебя совершенно неверная логика работы и то что тут происходит -- дичь.

DimaTrushin commented on First.cpp:
```diff
@@ -78,13 +117,36 @@ void test_back_propagation() {
+    Eigen::VectorXf y(4);
+    y << -1, 2, 1, -2.7;
+    ML::LossFunction LF;
+    //std::cout << LF.count_loss_function(x, y);
```
Вот это не должно работать с твоей сигнатурой count_loss_function. Тут должно быть что-то вроде
```cpp
LF.count_loss_function({x, y});
```
Однако, у тебя в целом не верно устроена LossFunction так что я бы совсем по-другому делал тесты.

DimaTrushin commented on First.cpp:
```diff
@@ -25,39 +28,74 @@ namespace ML {
     public:
         using Matrix = Eigen::MatrixXf;
         using Vector = Eigen::VectorXf;
+        Eigen::Rand::Vmt19937_64 urng{42};
```
Я бы сделал статические отдельный класс по созданию случайных матриц. И тут его надо сделать статическим, потому что он должен быть инициализирован один раз и не храниться в каждой копии твоего объекта. Например
```cpp
class Random {
using Matrix = Eigen::MatrixXf;
using Vector = Eigen::VectorXf;
public:
Random() = default;
Random(int seed) : rung{seed} {}

Matrix makeMatrix(int rows, int columns) {
  return norm_gen.template generate<Matrix>(rows, columns, urng);
}
Vector makeVector(int rows) {
  return norm_gen.template generate<Matrix>(rows, 1, urng);
}
private:
Eigen::Rand::Vmt19937_64 rung{42};
Eigen::Rand::NormalGen<float> norm_gen{0, 10};
};
```
Этот класс надо создать отдельно. После этого тут делаешь так
```cpp
private:
static Random Generator_;
public:
NetLayer(int rows, int columns) 
  : A_(Generator_::makeMatrix(rows, columns)),
    b_(Generator_::makeVector(rows)) {}

```

SobolevNick replied:
Сделано

DimaTrushin commented on First.cpp:
```diff
@@ -25,39 +28,74 @@ namespace ML {
         }
-        Matrix count_grad_A(Vector x, Vector u) const {
-            return sigma::evaluate1(A_ * x + b_) * u * x.transpose();
+        Matrix count_grad_A(const Vector &x, const Vector &u) const {
```
Я думаю, что нам понадобится метод который зависит не от x, а от z = A_ * x + b_, который мы будем хранить в памяти во время обучения. Но этот метод придется назвать как-нибудь по-другому, потому что по типу x и z  не отличаются.

SobolevNick replied:
Я правильно понимаю, что мы z всё равно получаем из функции?
Может, введём новую функцию, которая будет считать evaluate1(A_ * x + b_)

DimaTrushin replied:
Давай пока отложим эту часть оптимизации.

DimaTrushin reviewed (changes requested):
У тебя есть глобальная проблема с устройством нейросети и loss function. У тебя почему-то слои сети создаются и удаляются налету во время вычисления loss function. Я внес указания как и что надо поправить.

SobolevNick committed:
Update First.cpp

Я сделал функцию потерь, нейросеть пока не менял. 
У меня возник вопрос по использованию структуры, когда я хочу скормить в тест. 
Когда я вызываю TrainingDatum, он просто не видит её. 
Может, структуру вынести за класс?

DimaTrushin commented on First.cpp:
```diff
@@ -24,15 +23,33 @@ namespace ML {
-        NetLayer(int m, int n)
-            : A_(norm_gen.template generate<Matrix>(m, n, urng)), b_(norm_gen.template generate<Matrix>(m, 1, urng)) {
+        Random(unsigned long long seed) : rung{seed} {}
+        Matrix makeMatrix(int rows, int columns) {
```
Я понимаю, что это мой код был. Но теперь давай расскажу про одну тонкость. У тебя на вход функции подаются числа типа int. Они могут быть отрицательные и нулевые. Такие значения нам не подходят. То есть если ты вместо rows или columns поставишь отрицательное число, то это баг. Чтобы отлавливать такие баги обычно ставят специальную функцию, проверяющую, что условие выполняется. В качестве такой функции выступает assert. Тут в начале функции надо написать
```cpp
assert(rows > 0 && "rows must be positive!");
assert(columns > 0 && "columns must be positive!");
```
Функция assert в дебаг режиме проверяет условие внутри и если оно не выполнено, то роняет программу с сообщением об ошибке. В релизе (точнее, когда определен макрос NDEBUG) assert не генерирует никакого кода. Это один из стандартных способов проверки себя на вшивость.

DimaTrushin commented on First.cpp:
```diff
@@ -53,36 +70,40 @@ namespace ML {
-            for (int i = 0; i < x.size(); i++) {
-                sum = (Layer.predict(x[i].first) - x[i].second).dot(Layer.predict(x[i].first) - x[i].second) + sum;
+            for (int i = 0; i < z.size(); ++i) {
+                sum = (z[i] - data[i].Output).dot(z[i] - data[i].Output) + sum;
```
Используй
```cpp
m += expression;
```
вместо
```cpp
m = m + expression;
```

DimaTrushin commented on First.cpp:
```diff
@@ -53,36 +70,40 @@ namespace ML {
+        };
+        using TrainingData = std::vector<TrainingDatum>;
+        static float count_loss_function(const TrainingData &data, const std::vector<Vector> &z) {
             float sum = 0;
```
плохое название переменной sum. Имя ничего не говорит о хранимом в ней значении. Лучше тогда уж назвать result или loss.

DimaTrushin commented on First.cpp:
```diff
@@ -53,36 +70,40 @@ namespace ML {
-                sum = Layer.predict(x[i].first) - x[i].second + sum;
+        static Vector count_starting_gradient(const TrainingData &data, const std::vector<Vector> &z) {
+            Vector sum(z.size());
+            sum = Vector::Zero(z.size());
```
- Во-первых, ты можешь объединить эти строчки и написать
Vector sum= Vector::Zero(z.size());
- Во-вторых, я не понял, а какого у тебя размер sum совпадает с размером z? Размер z -- это же количество обучающих семплов. Что ты тут собираешься считать?
- Судя по всему sum -- это gradient. И ты хочешь, чтобы его размер был такой же как и размер всех элементов z (или как размер y). Я бы поставил тут assert(!z.empty()); и потом написал Vector gradient = Vector::Zero(z[0].size());

SobolevNick replied:
Да, конечно, мой косяк. 
Всё понятно. 
Я ещё добавил assert(!z.empty()); для предыдущего метода.

DimaTrushin commented on First.cpp:
```diff
@@ -53,36 +70,40 @@ namespace ML {
+        static Vector count_starting_gradient(const TrainingData &data, const std::vector<Vector> &z) {
+            Vector sum(z.size());
+            sum = Vector::Zero(z.size());
+            for (int i = 0; i < z.size(); i++) {
```
используй ++i вместо i++

SobolevNick replied:
Сделано.

DimaTrushin commented on First.cpp:
```diff
@@ -53,36 +70,40 @@ namespace ML {
         using Matrix = Eigen::MatrixXf;
         using Vector = Eigen::VectorXf;
-        NeuralNetwork(Vector x, Vector y) {
+        NeuralNetwork(const Vector x, const Vector y) {
```
Конструктор нейросети должен принимать на вход размеры слоев, а не векторы x и y. Векторы с обучающей выборкой должны подаваться на вход методу тренировки сети. Кроме того, передавать надо по константной ссылке const Vector& иначе у тебя все равно тут делается копирование.

SobolevNick replied:
Я это уже понял и написал, что нейросеть не трогал с того времени, когда было ничего не понятно.

SobolevNick committed:
Update First.cpp

DimaTrushin commented on First.cpp:
```diff
@@ -56,16 +57,32 @@ namespace ML {
             return Sigma::evaluate0(A_ * x + b_);
         }
         Matrix count_grad_A(const Vector &x, const Vector &u) const {
+            assert(x.size() == A_.cols() && "dimension x must be equal count columns A_!");
```
must be equal to the number of columns of A_

SobolevNick replied:
Ясно

DimaTrushin commented on First.cpp:
```diff
@@ -75,46 +92,82 @@ namespace ML {
-            return sum / z.size();
+            return result / z.size();
         }
         static Vector count_starting_gradient(const TrainingData &data, const std::vector<Vector> &z) {
```
Вот этот метод не верный. Ты его не сможешь потом в обучении использовать. Давай объясню проблему. Пусть у тебя есть сетка с двумя слоями
f : R^n -> R^m,  g : ^m -> R^k, L : R^k -> R.
Тогда сеть дает отображение g( f( x, a), b), где a, b -- параметры. А функция потерь считается так
1/n \sum_{i=1}^n L(g(f(x_i, a), b), y_i)
Тогда если ты хочешь посчитать градиент по b, то ты получаешь
1/n \sum_{i=1}^n d/dz L(g(f(x_i, a), b), y_i) d/db g(f(x_i, a), b)
То есть тебе недостаточно знать среднее значение от градиентов
d/dz L(g(f(x_i, a), b), y_i)
тебе надо уметь каждое слагаемое умножать на соответствующий якобиан и только потом усреднять. Потому тебе нужен метод, который считает только одно слагаемое. И только потом усреднять.

SobolevNick replied:
То есть этот метод принимает на вход одну тройку x, y, z?

SobolevNick replied:
А мне на якобиан умножать надо здесь же будет, я правильно понимаю?

DimaTrushin replied:
Чтобы не было ошибок. Я бы для начала сделал методы от одной обучающей пары или в данном случае от одной тройки x, y, z. Просто иначе, я боюсь, будет много ошибок и тупиковая ситуация.

DimaTrushin replied:
Про якобиан не очень понял. Тут тебе надо лишь градиент Loss функции на одной тройке посчитать и вернуть как вектор. Якобиан возникнет, если ты захочешь передавать эти данные дальше по слоям.

SobolevNick replied:
Мне кажется, лучше брать не тройку x, y, z, а пару x, z, так как в случае для расчёта начального градиента для второго слоя x - это z1[i], y - data.Output[i], z - z2[i]. Придётся создавать новую структуру, запихивая туда z1 и data.Output. 
Это ещё ладно, но если считать для первого слоя, то там мы вообще не знаем y. 
Что скажете?

SobolevNick replied:
Бред написал, читать не надо, видимо, не проснулся.

DimaTrushin commented on First.cpp:
```diff
@@ -75,46 +92,82 @@ namespace ML {
+            std::vector<Vector> z2;
+            Vector u1;
+            Vector u2;
+            Vector z1_mean = Vector::Zero(z1[0].size());
```
У тебя z1 является пустым вектором, в нем нет элемента z1[0]. Потому тут программа упадет.

SobolevNick replied:
Мне надо перенести эту строчку вниз, после того как z1 заполню.

SobolevNick replied:
Его заполнять не надо - я понял)

DimaTrushin commented on First.cpp:
```diff
@@ -75,46 +92,82 @@ namespace ML {
+            Vector u1;
+            Vector u2;
+            Vector z1_mean = Vector::Zero(z1[0].size());
+            Vector x1_mean = Vector::Zero(data[0].Input.size());
```
сделай в начале train проверку, что data не пусто. Я бы сделал так
```cpp
if(data.size() == 0)
  return;
```

SobolevNick replied:
Хорошо. 
Я просто именно этот класс ещё на вшивость не проверял, хочу чтобы он был правильный при правильных данных и только потом уже ругаться, если данные плохие

DimaTrushin commented on First.cpp:
```diff
@@ -75,46 +92,82 @@ namespace ML {
+            Vector x1_mean = Vector::Zero(data[0].Input.size());
+            for (int i = 0; i < data.size(); ++i) {
+                z1[i] = NL1.predict(data[i].Input);
+                z2[i] = NL2.predict(data[i].Input);
```
Здесь должно быть
```cpp
z2[i] = NL2.predict(z1[i]);
```
потому что ты считаешь значения второго слоя на выходе значений первого слоя.

SobolevNick replied:
Да, конечно. 
Исправлю.

DimaTrushin commented on First.cpp:
```diff
@@ -75,46 +92,82 @@ namespace ML {
+                z1[i] = NL1.predict(data[i].Input);
+                z2[i] = NL2.predict(data[i].Input);
+            }
+            u2 = LF.count_starting_gradient(data, z2);
```
Начиная отсюда, тебе надо переписать код следующим образом.
Ты берешь одну обучающую пару data[i]. Для нее ты рассчитал промежуточные значения z1[i] и z2[i]. После этого ты должен посчитать начальный градиент u2[i]. После этого для каждого i ты считаешь gradA2[i] и gradb2[i] для второго слоя. И считаешь пробрасываемый градиент u1[i]. После чего ты усредняешь их по i и сдвигаешь A_ и b_ второго слоя. Потом ты с помощью u1[i] считаешь gradA1[i] и gradb1[i] для первого слоя. И потом уже усредняешь их и только потом сдвигаешь A_ и b_ первого слоя.

SobolevNick replied:
Так, это вроде понятно. 
z1_mean и x1_mean вообще не надо

DimaTrushin reviewed (changes requested):
Качество кода стало хорошим. Но есть содержательная проблема с методом train. Я постарался расписать, что и как надо поправить. Если будут вопросы -- пиши.

DimaTrushin reviewed (changes requested):
Из того, что хочется увидеть в изменениях
1) Сделай Loss Function только от одной TrainingDatum.
2) Поправь код для обучения Нейросети. Сделай обучение за раз на одной паре, как я писал.

SobolevNick committed:
Update First.cpp

DimaTrushin commented on First.cpp:
```diff
@@ -112,15 +112,11 @@ namespace ML {
+            assert(z.size() != 0);
+            assert(z.size() == data.Output.size() && "dimension z must be equal dimension data.Output");
+            Vector gradient = z - data.Output;
+            return 2 * gradient;
```
Перепиши это так
```cpp
Vector gradient = 2* (z - data.Output);
return gradient;
```
В этом случае гарантировано сработает NRVO. То есть объект gradient будет сразу сконструирован по адресу, куда возвращает функция. Плюс так идейно правильнее, так как градиент равен удвоенной разности. Нет смысла выражение разбивать на части.

DimaTrushin commented on First.cpp:
```diff
@@ -136,28 +132,30 @@ namespace ML {
-            Vector x1_mean = Vector::Zero(data[0].Input.size());
+            std::vector<Vector> u1;
+            std::vector<Vector> u2;
+            Matrix grad_A2;
```
У меня вопрос (это надо тестить). Тут ты не инициализируешь градиент. А потом используешь grad_A2 += что-то. По хорошему у grad_A2 должно быть нулевое стартовое значение. Надо проверить, что это действительно правильно работает или тут надо что-то вроде Matrix grad_A2 = ZeroMatrix(rows, columns) или что-то в этом духе (это приблизительный код я наизусть не помню, как нулевую матрицу в eigen задавать).

DimaTrushin commented on First.cpp:
```diff
@@ -136,28 +132,30 @@ namespace ML {
         void train(const TrainingData &data) {
```
Вот эта функция выглядит хорошо.

DimaTrushin reviewed (changes requested):
То что я вижу мне нравится. train теперь должна быть правильной. Сделал мелкие замечания.

SobolevNick committed:
Update First.cpp

SobolevNick committed:
Update First.cpp

SobolevNick committed:
Update First.cpp

DimaTrushin commented on First.cpp:
```diff
@@ -58,29 +58,38 @@ namespace ML {
             assert(A.cols() == A_.cols() && "dimensions of matrices must be same");
             assert(A.rows() == A_.rows() && "dimensions of matrices must be same");
-            A_ -= A;
+            A_ = A_ - A;
```
Зачем сделано это изменение от -= к =?

SobolevNick committed:
Update First.cpp

DimaTrushin commented on First.cpp:
```diff
@@ -9,19 +10,46 @@ namespace ML {
+        static Vector evaluate0(const Vector &x) {
+            Vector y(x.size());
+            for (int i = 0; i < x.size(); ++i) {
+                y[i] = (x[i] > 0) ? x[i] : 0;
```
Тут можно написать
```cpp
y[i] = evaluate0(x[i]);
```

DimaTrushin commented on First.cpp:
```diff
@@ -1,81 +1,351 @@
+                u2[i] = LF.count_starting_gradient(data[i], z2[i]);
+                u1[i] = NL2.count_grad_x(z1[i], u2[i]);
+            }
+            int n = 1;
```
n и j -- очень плохие названия переменных, которые ни о чем не говорят.

DimaTrushin commented on First.cpp:
```diff
@@ -1,81 +1,351 @@
+                v_A1.resize(j + 1);
+                v_b1.resize(j + 1);
+                if (j != 0) {
+                    v_A2[j] = 0.9 * v_A2[j - 1] + grad_A2 / n;
```
В связи с чем такие сложные формулы?
Мне кажется, что увеличение n на единичку с каждым шагом -- слишком сильно зануляет последний градиент. У тебя основной вклад будет от первых градиентов.

DimaTrushin commented on First.cpp:
```diff
@@ -1,81 +1,351 @@
+                grad_b2 = grad_b2 / data.size();
+                grad_A1 = grad_A1 / data.size();
+                grad_b1 = grad_b1 / data.size();
+                v_A2[0] = grad_A2;
```
Вот эти обновления будут делаться на каждой итерации цикла. Судя по всему их надо засунуть вниз у тебя условие, если j = 0 то делаешь эти строчки, если j != 0, то делаешь другие.

DimaTrushin commented on First.cpp:
```diff
@@ -1,81 +1,351 @@
+                v_b2[0] = grad_b2;
+                v_A1[0] = grad_A1;
+                v_b1[0] = grad_b1;
+                v_A2.resize(j + 1);
```
Это достаточно дорогая операция. Лучше v_A2 определить вне цикла и там сделать reserve(максимальный размер). Тогда у тебя не будет реаллокации в цикле.

DimaTrushin reviewed (changes requested):
Набросал минимальные комментарии.
